library/pkcs7.c

```diff
@@ -507,34 +573,62 @@ int mbedtls_pkcs7_signed_data_verify( mbedtls_pkcs7 *pkcs7,
                                       size_t datalen )
 {
 
-    int ret;
+    int ret = MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
     unsigned char *hash;
     mbedtls_pk_context pk_cxt = cert->pk;
     const mbedtls_md_info_t *md_info;
     mbedtls_md_type_t md_alg;
+    mbedtls_pkcs7_signer_info *signer;
 
-    ret = mbedtls_oid_get_md_alg( &pkcs7->signed_data.digest_alg_identifiers, &md_alg );
-    if( ret != 0 )
+    if( pkcs7->signed_data.no_of_signers == 0 )
         return( MBEDTLS_ERR_PKCS7_VERIFY_FAIL );
 
-    md_info = mbedtls_md_info_from_type( md_alg );
 
-    hash = mbedtls_calloc( mbedtls_md_get_size( md_info ), 1 );
-    if( hash == NULL ) {
-        return( MBEDTLS_ERR_PKCS7_ALLOC_FAILED );
-    }
-
-    ret = mbedtls_md( md_info, data, datalen, hash );
-    if( ret != 0 )
+    /*
+     * Potential TODOs
+     * Currently we iterate over all signers and return success if any of them
+     * verify.
+     *
+     * However, we could make this better by checking against the certificate's
+     * identification and SignerIdentifier fields first. That would also allow
+     * us to distinguish between 'no signature for key' and 'signature for key
+     * failed to validate'.
+     *
+     * We could also cache hashes by md, so if there are several sigs all using
+     * the same algo we don't recalculate the hash each time.
+     */
+    signer = &pkcs7->signed_data.signers;
+    while( signer )
     {
+        ret = mbedtls_oid_get_md_alg( &signer->alg_identifier, &md_alg );
+        if( ret != 0 )
+            return( MBEDTLS_ERR_PKCS7_VERIFY_FAIL );
+
+        md_info = mbedtls_md_info_from_type( md_alg );
```

Comment:
`pkcs7_get_digest_algorithm_set()` in the latest version of the PR only supports a single element, so either all the signatures will use the same algorithm, or the parsing will error out before getting here.

Reply:
Kind of yes and no.
Technically, it IS possible but whether or not it is a valid PKCS7 data structure, I am unsure. Let me try to elaborate, in RFC 2315, the SignedData portion of a PKCS7 is defined as:
```
SignedData
 ::= SEQUENCE {
     version Version,
     digestAlgorithms DigestAlgorithmIdentifiers,
     contentInfo ContentInfo,
     certificates
        [0] IMPLICIT ExtendedCertificatesAndCertificates
          OPTIONAL,
     crls
       [1] IMPLICIT CertificateRevocationLists OPTIONAL,
     signerInfos SignerInfos }
```
 In `pkcs7_get_signed_data` we call `pkcs7_get_digest_algorithm_set` to assign a value to the mbedtls representation of `digestAlgorithms` . It is important to note the the RFC state this value "may be any number of elements, including zero", with its main purpose to help with single pass verification, in mbedtls world, you are correct, we have a strong enforcement of exactly one (depending on feedback, I could see how this would want to be handled differently).
 
 But, referring back to your comment above, the value in `signed_data->signers->alg_identifier` is a different value than `signed_data->digest_alg_identifiers` (ideally related but in terms of memory different). `signers->alg_identifier` is the mbedtls representation of the RFC value `digestAlgorithm` in the `SignerInfo` structure:
 ```
 SignerInfo ::= SEQUENCE {
     version Version,
     issuerAndSerialNumber IssuerAndSerialNumber,
     digestAlgorithm DigestAlgorithmIdentifier,
     authenticatedAttributes
       [0] IMPLICIT Attributes OPTIONAL,
     digestEncryptionAlgorithm
       DigestEncryptionAlgorithmIdentifier,
     encryptedDigest EncryptedDigest,
     unauthenticatedAttributes
       [1] IMPLICIT Attributes OPTIONAL }
 ```
 Simplifying the pkcs7 data structure down we have:
 ```
 pkcs7:
     SignedData:
         - digestAlgorithms [ ]
         - SignerInfos [ ]
              - SignerInfo[1]:
                  - digestAlgorithm
              - SignerInfo[2]:
                  - digestAlgortihm
             etc
```
The `digestAlgorithm` value we parse correctly and one is assigned to every signer. When we are verifying, we loop over this value instead of the higher level `digestAlgorithms` because `digestAlgorithm` is more reliable. Please let me know if this made things more/less confusing